source/credentials_provider_imds.c

```diff
@@ -565,86 +354,418 @@ static int s_make_imds_http_query(
         .request = request,
     };
 
-    struct aws_http_stream *stream =
-        impl->function_table->aws_http_connection_make_request(imds_user_data->connection, &request_options);
+    stream = impl->function_table->aws_http_connection_make_request(imds_user_data->connection, &request_options);
 
-    return stream == NULL ? AWS_OP_ERR : AWS_OP_SUCCESS;
+    if (!stream) {
+        goto on_error;
+    }
 
-on_error:
+    if (impl->function_table->aws_http_stream_activate(stream)) {
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
 
+on_error:
+    impl->function_table->aws_http_stream_release(stream);
```

Comment:
Should we set the error state here on the user data?

Reply:
this isn't a state function. Only state functions are allowed to mutate state.